server/lorax_server/models/custom_modeling/flash_qwen2_modeling.py

```diff
@@ -496,7 +495,64 @@ def forward(
             prefill_cache_indices,
             adapter_data,
         )
+
         if lm_head_indices is not None:
             hidden_states = hidden_states[lm_head_indices]
         logits, speculative_logits = self.lm_head(hidden_states, adapter_data)
         return logits, speculative_logits
+
+
+class FlashQwen2ForEmbeddings(torch.nn.Module):
+    def __init__(self, config, weights):
+        super().__init__()
+        self.config = config
+
+        self.model = FlashQwen2Model(config, weights)
+        self.max_past = config.sliding_window
+        self.output_weight = weights.get_tensor("linear.weight")
+        self.output_bias = weights.get_tensor("linear.bias")
+        # To satisfy the parent class interface
+        # TODO: fix
+        self.lm_head = None
+
+    def forward(
+        self,
+        input_ids: torch.Tensor,
+        position_ids: torch.Tensor,
+        cu_seqlen_prefill: Optional[torch.Tensor],
+        kv_cache: List[Tuple[torch.Tensor, torch.Tensor]],
+        block_tables: torch.Tensor,
+        slots: torch.Tensor,
+        input_lengths: torch.Tensor,
+        max_s: int,
+        adapter_data: AdapterBatchData,
+        prefill_cache_indices: Optional[torch.Tensor] = None,
+        lm_head_indices: Optional[torch.Tensor] = None,
+    ) -> Tuple[torch.Tensor, Optional[torch.Tensor]]:
+        if prefill_cache_indices is not None:
+            # Slots also need to be sliced as it has the same size as the whole kv tensor
+            slots = slots[prefill_cache_indices]
+        elif self.max_past is not None:
+            # Clamp in decode mode as paged attention requires clamped values whereas the flash attention
+            # kernel requires the true values
+            max_s = min(self.max_past, max_s)
+            input_lengths = torch.clamp(input_lengths, max=self.max_past)
+
+        hidden_states = self.model(
+            input_ids,
+            position_ids,
+            cu_seqlen_prefill,
+            kv_cache,
+            block_tables,
+            slots,
+            input_lengths,
+            max_s,
+            prefill_cache_indices,
+            adapter_data,
+        )
+        batch_size = hidden_states.shape[0] // max_s
+        hidden_states = hidden_states.reshape(batch_size, max_s, -1)
+        mean_hidden_states = hidden_states.mean(1)
```

Comment:
Is this taking the mean along the sequence dimension?

Reply:
yes